### PR TITLE
FIX: do not let no-op monkey patches to renderer leak out

### LINKED ISSUE #17542: Matplotlib 3.2.1 savefig empty image when fig size matches data size exactly
### Bug report

**Bug summary**

Saving a figure to an image file with the same pixel dimensions as data loaded with imshow does not work in 3.2.1, producing an empty image. If the dimensions are changed by even one pixel, normal behavior returns.

This bug appeared in 3.2.1 and I have confirmed that it does not occur with 3.1.1.

I noticed in the release notes for 3.2.0 that some changes were made to the default image interpolation; perhaps an unhandled condition was introduced here when no interpolation is required. However, I note that changing the interpolation mode had no effect on this bug.  

**Code for reproduction**

```python
import numpy as np
from matplotlib import pylab as plt

x_size, y_size = (100, 100)
fig = plt.figure(frameon=False, dpi=100, figsize=(y_size/100, x_size/100))
ax = plt.Axes(fig, [0., 0., 1., 1.])
fig.add_axes(ax)
ax.set_axis_off()
ax.get_xaxis().set_visible(False)
ax.get_yaxis().set_visible(False)

data = np.random.rand(100, 100)
ax.imshow(data)
fig.savefig('/tmp/image.png', bbox_inches='tight', pad_inches=0)

# In matplotlib 3.2.1, empty file with size 100x100 pixels output
# if random data shape is changed to something other than 100x100, output
# works correctly.

# In matplotlib 3.1.1, file output is 100x100 with data shown correctly.
```

**Actual outcome**
bug_image

**Expected outcome**

working_image

Produced with matplotlib 3.1.1

**Matplotlib version**
  * Operating system: Linux version 5.6.13-arch1-1
  * Matplotlib version: 3.2.1
  * Matplotlib backend (`print(matplotlib.get_backend())`): TkAgg
  * Python version: 3.8.3
  * Other libraries: numpy 1.18.4

Default system install on Arch linux package `python-matplotlib-3.2.1-1-x86_64`

### PR DESCRIPTION
**Pull request by tacaswell**
## PR Summary

closes #17542

This issue is that in some cases the no-op monkey patched renderer was getting cached deep in the backend code and the no-op functions were getting used for the actual save which resulted in empty figures.  This moves the monkey patching to the call sites so we can use the `_setattr_cm` to manage this.

## PR Checklist

- [x] Has Pytest style unit tests
- [x] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [NA] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way (internal api only)

### CODE REVIEW

**Comment by tacaswell** [comment]
I'm 50/50 on moving it to the base class. 

On one hand I agree it is much neater.

On the other hand, how strongly do we require the renderers to actually be sub-classes?  I would say that adding a new method to the semi-public API of the renderer is something we should not do in a patch release and we probably should backport this to 3.2.2 to fix the empty figure bug.

**Comment by anntzer** [comment]
We basically do per https://github.com/matplotlib/matplotlib/pull/17159.
If you really want to be super safe we can add an isinstance check and if the renderer subclass is not a RendererBase, then patch the method onto it, or even just make that a free private function in backend_bases (but then subclasses can't override it).

**Comment by tacaswell** [comment]
Will add helper function.

**Comment by tacaswell** [comment]
I made it forgiving for now (to backport to 3.2.x) but I'll open a follow-on PR to make it warn for 3.3.

File: lib/matplotlib/backend_bases.py [hunk 1]
```diff
@@ -1502,15 +1520,14 @@ def __init__(self, name, canvas, key, x=0, y=0, guiEvent=None):
...
     """
     Get the renderer that would be used to save a `~.Figure`, and cache it on
     the figure.
 
-    If *draw_disabled* is True, additionally replace drawing methods on
-    *renderer* by no-ops.  This is used by the tight-bbox-saving renderer,
-    which needs to walk through the artist tree to compute the tight-bbox, but
-    for which the output file may be closed early.
+    If you need a renderer without any active draw methods use
+    cbook._setattr_cm to temporary patch them out at your call site.
```
Comment (anntzer) [inline lib/matplotlib/backend_bases.py:1529]: update this?

**Pull request merged**